### ISSUE TITLE
Webhook should support basic auth

### ISSUE THREAD

**Issue by st2stanley** [StackStorm/st2#2328]
In some cases, a user will need to expose StackStorm to the public world. Currently, we support token authorization, but not all services will allow header customization. The best available is just the URI. (for example: SNS). It would be fantastic if the same API key could ​*also*​ be used in basic-auth with the webhook sensor.

This was reported by @tt in #stackstorm-community

**Comment by Kami** [comment]
auth token and api key can already both be provided as query parameter as
part of the url

**Comment by jfryman** [comment]
@Kami I'm debating this with @dzimine right now. Having in the URI as a query parameter *could* be a security risk, even though it's wrapped in SSL. Versus, basic auth will do what we want and send as a header as well.

**Comment by jfryman** [comment]
I don't think this is as high priority as I thought... I did overlook the query param. So, not immediate rush, but something to think on.

**Comment by Kami** [comment]
> @Kami I'm debating this with @dzimine right now. Having in the URI as a query parameter could be a security risk, even though it's wrapped in SSL. Versus, basic auth will do what we want and send as a header as well.

Yeah to add some context - the support for sending token via query param was originally because of the WebUI. That's also why it wasn't documented - it was only there for WebUI.

Having said that, I do think sending it in query params carries some risk, but it's most portable way and most of the services I'm aware of either include sending secret in the path or query param. As you have said - not all the services support sending custom headers, same goes for basic auth credentials in the URL syntax (https://foo:bar@foo.com), not many services support that either so query param is the lowest common denominator.

In short - if we all agree with that we should document that feature :)

**Comment by lakshmi-kannan** [comment]
+1 to documenting it.

**Comment by Kami** [comment]
@lakshmi-kannan Yeah, I will submit a PR there it's documented (I will of course also include potential security risk disclaimer).

**Comment by Kami** [comment]
I merged documentation so I will consider this as complete for now.

I don't think query parameters are much less secure than headers and basic auth doesn't really solve the compatibility issues since a lot of third party services don't allow you to specify custom basic auth credentials or don't support specifying credentials in the URL so it's really not much better than the existing headers approach.

So from the integration / compatibility perspective providing token in query param or path is the best approach - most other services use this approach as well (think Slack webhooks, etc.)
